File: public/scripts/download_images.php
<?php

/**
 * Image Downloader Script for PetTrack
 * 
 * This script extracts all image URLs from the home page and downloads them
 * to a local directory structure that mirrors the original paths.
 */

// Configuration
$saveDirectory = __DIR__ . '/../images/downloaded/';
$logFile = __DIR__ . '/image_download_log.txt';

// Create the save directory if it doesn't exist
if (!file_exists($saveDirectory)) {
    mkdir($saveDirectory, 0755, true);
}

// Initialize log file
$log = fopen($logFile, 'w');
fwrite($log, "Image Download Log - " . date('Y-m-d H:i:s') . "\n\n");

// Function to extract image URLs from HTML content
function extractImageUrls($html) {
    $imageUrls = [];
    
    // Extract background images from inline styles
    preg_match_all('/background-image:\s*url\([\'"]([^\'"]+)[\'"]\)/', $html, $bgMatches);
    if (!empty($bgMatches[1])) {
        $imageUrls = array_merge($imageUrls, $bgMatches[1]);
    }
    
    // Extract image URLs from img tags
    preg_match_all('/<img[^>]+src=[\'"]([^\'"]+)[\'"][^>]*>/', $html, $imgMatches);
    if (!empty($imgMatches[1])) {
        $imageUrls = array_merge($imageUrls, $imgMatches[1]);
    }
    
    // Extract image URLs from PHP arrays (specific to this template)
    preg_match_all('/\[\'img\'\] => \'([^\']+)\'/', $html, $phpArrayMatches);
    if (!empty($phpArrayMatches[1])) {
        $imageUrls = array_merge($imageUrls, $phpArrayMatches[1]);
    }
    
    return array_unique($imageUrls);
}

// Function to download an image
function downloadImage($url, $saveDirectory, $log) {
    // Parse URL to get the path
    $parsedUrl = parse_url($url);
    $path = isset($parsedUrl['path']) ? $parsedUrl['path'] : '';
    
    // Remove query parameters if present
    $path = preg_replace('/\?.*/', '', $path);
    
    // Get the filename from the path
    $filename = basename($path);
    
    // Create a directory structure based on the domain
    $domain = isset($parsedUrl['host']) ? $parsedUrl['host'] : 'unknown';
    $domainDir = $saveDirectory . $domain . '/';
    
    if (!file_exists($domainDir)) {
        mkdir($domainDir, 0755, true);
    }
    
    // Generate a unique filename if needed
    $savePath = $domainDir . $filename;
    $counter = 1;
    while (file_exists($savePath)) {
        $info = pathinfo($filename);
        $savePath = $domainDir . $info['filename'] . '_' . $counter . '.' . (isset($info['extension']) ? $info['extension'] : '');
        $counter++;
    }
    
    // Download the image
    try {
        $imageContent = @file_get_contents($url);
        if ($imageContent === false) {
            fwrite($log, "ERROR: Could not download $url\n");
            return false;
        }
        
        file_put_contents($savePath, $imageContent);
        fwrite($log, "SUCCESS: Downloaded $url to $savePath\n");
        return true;
    } catch (Exception $e) {
        fwrite($log, "ERROR: Exception while downloading $url: " . $e->getMessage() . "\n");
        return false;
    }
}

// Get the home page content
$homePagePath = __DIR__ . '/../../resources/views/home.blade.php';
$homePageContent = file_get_contents($homePagePath);

if ($homePageContent === false) {
    fwrite($log, "ERROR: Could not read home page template\n");
    die("Could not read home page template");
}

// Extract image URLs
$imageUrls = extractImageUrls($homePageContent);

// Download each image
$totalImages = count($imageUrls);
$successCount = 0;
$failCount = 0;

fwrite($log, "Found $totalImages unique image URLs\n\n");

echo "<!DOCTYPE html>
<html>
<head>
    <title>Image Download Progress</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            max-width: 800px;
            margin: 0 auto;
            padding: 20px;
        }
        .progress-container {
            width: 100%;
            background-color: #f3f3f3;
            border-radius: 5px;
            margin: 20px 0;
        }
        .progress-bar {
            height: 30px;
            background-color: #4CAF50;
            border-radius: 5px;
            text-align: center;
            line-height: 30px;
            color: white;
        }
        table {
            width: 100%;
            border-collapse: collapse;
        }
        th, td {
            padding: 8px;
            text-align: left;
            border-bottom: 1px solid #ddd;
        }
        th {
            background-color: #f2f2f2;
        }
        .success {
            color: green;
        }
        .error {
            color: red;
        }
    </style>
</head>
<body>
    <h1>Image Download Progress</h1>
    <div class='progress-container'>
        <div class='progress-bar' id='progressBar' style='width: 0%'>0%</div>
    </div>
    <p id='status'>Downloading <span id='current'>0</span> of <span id='total'>{$totalImages}</span> images...</p>
    
    <h2>Image List</h2>
    <table>
        <thead>
            <tr>
                <th>#</th>
                <th>Image URL</th>
                <th>Status</th>
            </tr>
        </thead>
        <tbody>";

// Process each image and update the UI
foreach ($imageUrls as $index => $url) {
    $current = $index + 1;
    $percent = round(($current / $totalImages) * 100);
    
    echo "<tr>
            <td>{$current}</td>
            <td>{$url}</td>";
    
    $result = downloadImage($url, $saveDirectory, $log);
    
    if ($result) {
        $successCount++;
        echo "<td class='success'>Success</td>";
    } else {
        $failCount++;
        echo "<td class='error'>Failed</td>";
    }
    
    echo "</tr>";
    
    // Flush output to show progress
    echo "<script>
        document.getElementById('progressBar').style.width = '{$percent}%';
        document.getElementById('progressBar').innerText = '{$percent}%';
        document.getElementById('current').innerText = '{$current}';
    </script>";
    
    flush();
    ob_flush();
    
    // Small delay to avoid overwhelming the server
    usleep(100000); // 100ms
}

// Summary
fwrite($log, "\nSummary:\n");
fwrite($log, "Total images: $totalImages\n");
fwrite($log, "Successfully downloaded: $successCount\n");
fwrite($log, "Failed: $failCount\n");
fclose($log);

echo "</tbody>
    </table>
    
    <h2>Summary</h2>
    <p>Total images: {$totalImages}</p>
    <p>Successfully downloaded: {$successCount}</p>
    <p>Failed: {$failCount}</p>
    
    <p>A detailed log has been saved to: {$logFile}</p>
    
    <h2>Next Steps</h2>
    <p>To implement these images locally in your templates:</p>
    <ol>
        <li>Replace external URLs in your templates with local paths</li>
        <li>For example, replace <code>https://images.pexels.com/photos/1108099/pexels-photo-1108099.jpeg</code> with <code>/images/downloaded/images.pexels.com/photos/1108099/pexels-photo-1108099.jpeg</code></li>
        <li>You can use the following PHP script to automate this replacement:</li>
    </ol>
    
    <pre>
&lt;?php
// Script to replace image URLs in template files
$templateFile = '/path/to/template.blade.php';
$content = file_get_contents($templateFile);

// Define replacements (original URL => local path)
$replacements = [
    'https://images.pexels.com/photos/1108099/pexels-photo-1108099.jpeg' => '/images/downloaded/images.pexels.com/photos/1108099/pexels-photo-1108099.jpeg',
    // Add more replacements as needed
];

// Perform replacements
foreach ($replacements as $original => $replacement) {
    $content = str_replace($original, $replacement, $content);
}

// Save the updated content
file_put_contents($templateFile, $content);
echo 'Template updated successfully!';
?&gt;
    </pre>
    
    <p><a href='/'>Return to Home</a></p>
</body>
</html>";

File: download_images.py
#!/usr/bin/env python3
"""
PetTrack Image Downloader

This script extracts all image URLs from the home page and downloads them
to a local directory structure for local implementation.
"""

import os
import re
import sys
import time
import urllib.parse
import urllib.request
import shutil
from datetime import datetime
from pathlib import Path

# Configuration
SCRIPT_DIR = Path(__file__).parent
HOME_PAGE_PATH = SCRIPT_DIR / "resources" / "views" / "home.blade.php"
SAVE_DIRECTORY = SCRIPT_DIR / "public" / "images" / "downloaded"
LOG_FILE = SCRIPT_DIR / "image_download_log.txt"

# ANSI color codes for terminal output
class Colors:
    HEADER = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

def print_header(text):
    """Print a formatted header."""
    print(f"\n{Colors.HEADER}{Colors.BOLD}=== {text} ==={Colors.ENDC}\n")

def print_success(text):
    """Print a success message."""
    print(f"{Colors.GREEN}✓ {text}{Colors.ENDC}")

def print_error(text):
    """Print an error message."""
    print(f"{Colors.FAIL}✗ {text}{Colors.ENDC}")

def print_info(text):
    """Print an info message."""
    print(f"{Colors.CYAN}ℹ {text}{Colors.ENDC}")

def print_progress(current, total, prefix='', suffix='', length=50):
    """Print a progress bar."""
    percent = float(current) * 100 / total
    filled_length = int(length * current // total)
    bar = '█' * filled_length + '░' * (length - filled_length)
    sys.stdout.write(f'\r{prefix} |{bar}| {percent:.1f}% {suffix}')
    sys.stdout.flush()
    if current == total:
        print()

def extract_image_urls(html_content):
    """Extract all image URLs from HTML content."""
    image_urls = []
    
    # Extract background images from inline styles
    bg_pattern = r'background-image:\s*url\([\'"]([^\'"]+)[\'"]\)'
    bg_matches = re.findall(bg_pattern, html_content)
    if bg_matches:
        image_urls.extend(bg_matches)
    
    # Extract image URLs from img tags
    img_pattern = r'<img[^>]+src=[\'"]([^\'"]+)[\'"]'
    img_matches = re.findall(img_pattern, html_content)
    if img_matches:
        image_urls.extend(img_matches)
    
    # Extract image URLs from PHP arrays (specific to this template)
    php_array_pattern = r'\[\'img\'\] => \'([^\']+)\''
    php_array_matches = re.findall(php_array_pattern, html_content)
    if php_array_matches:
        image_urls.extend(php_array_matches)
    
    # Filter out non-HTTP URLs and remove duplicates
    http_urls = [url for url in image_urls if url.startswith('http')]
    return list(set(http_urls))

def download_image(url, save_directory, log_file):
    """Download an image and return the local path."""
    try:
        # Parse URL to get the path
        parsed_url = urllib.parse.urlparse(url)
        path = parsed_url.path
        
        # Remove query parameters if present
        path = path.split('?')[0]
        
        # Get the filename from the path
        filename = os.path.basename(path)
        
        # Create a directory structure based on the domain
        domain = parsed_url.netloc
        domain_dir = save_directory / domain
        domain_dir.mkdir(parents=True, exist_ok=True)
        
        # Generate a unique filename if needed
        save_path = domain_dir / filename
        relative_path = f"/images/downloaded/{domain}/{filename}"
        
        counter = 1
        while save_path.exists():
            name, ext = os.path.splitext(filename)
            new_filename = f"{name}_{counter}{ext}"
            save_path = domain_dir / new_filename
            relative_path = f"/images/downloaded/{domain}/{new_filename}"
            counter += 1
        
        # Download the image
        with urllib.request.urlopen(url, timeout=10) as response, open(save_path, 'wb') as out_file:
            shutil.copyfileobj(response, out_file)
        
        log_file.write(f"SUCCESS: Downloaded {url} to {save_path}\n")
        return relative_path
    
    except Exception as e:
        log_file.write(f"ERROR: Could not download {url}: {str(e)}\n")
        return None

def main():
    """Main function to download images."""
    print_header("PetTrack Image Downloader")
    
    # Create save directory if it doesn't exist
    SAVE_DIRECTORY.mkdir(parents=True, exist_ok=True)
    
    # Initialize log file
    with open(LOG_FILE, 'w') as log:
        log.write(f"Image Download Log - {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n\n")
        
        try:
            # Read the home page content
            print_info(f"Reading home page from {HOME_PAGE_PATH}")
            with open(HOME_PAGE_PATH, 'r', encoding='utf-8') as file:
                home_page_content = file.read()
            
            # Extract image URLs
            print_info("Extracting image URLs...")
            image_urls = extract_image_urls(home_page_content)
            total_images = len(image_urls)
            
            if total_images == 0:
                print_error("No image URLs found in the home page.")
                log.write("No image URLs found in the home page.\n")
                return
            
            print_success(f"Found {total_images} unique image URLs")
            log.write(f"Found {total_images} unique image URLs\n\n")
            
            # Download each image
            print_header("Downloading Images")
            success_count = 0
            fail_count = 0
            replacements = {}
            
            for i, url in enumerate(image_urls):
                print_progress(i, total_images, prefix='Progress:', suffix=f'({i}/{total_images})')
                print(f"\nDownloading: {url}")
                
                local_path = download_image(url, SAVE_DIRECTORY, log)
                
                if local_path:
                    success_count += 1
                    replacements[url] = local_path
                    print_success(f"Saved to {local_path}")
                else:
                    fail_count += 1
                    print_error(f"Failed to download {url}")
                
                # Small delay to avoid overwhelming servers
                time.sleep(0.1)
            
            print_progress(total_images, total_images, prefix='Progress:', suffix=f'({total_images}/{total_images})')
            
            # Summary
            print_header("Summary")
            print(f"Total images: {total_images}")
            print(f"Successfully downloaded: {success_count}")
            print(f"Failed: {fail_count}")
            
            log.write("\nSummary:\n")
            log.write(f"Total images: {total_images}\n")
            log.write(f"Successfully downloaded: {success_count}\n")
            log.write(f"Failed: {fail_count}\n")
            
            # Generate replacement guide
            if success_count > 0:
                print_header("Replacement Guide")
                print("To implement these images locally in your templates, replace the URLs as follows:")
                print("\nExample replacements:")
                
                # Show a few examples
                for i, (url, local_path) in enumerate(list(replacements.items())[:3]):
                    print(f"\n{i+1}. Original: {url}")
                    print(f"   Replace with: {{ asset('{local_path}') }}")
                
                # Create a replacement guide file
                guide_path = SCRIPT_DIR / "image_replacement_guide.txt"
                with open(guide_path, 'w') as guide:
                    guide.write("Image Replacement Guide\n")
                    guide.write("=====================\n\n")
                    guide.write("Replace the following URLs in your templates:\n\n")
                    
                    for url, local_path in replacements.items():
                        guide.write(f"Original: {url}\n")
                        guide.write(f"Replace with: {{ asset('{local_path}') }}\n\n")
                
                print(f"\nA complete replacement guide has been saved to {guide_path}")
            
            print(f"\nA detailed log has been saved to {LOG_FILE}")
            
        except Exception as e:
            print_error(f"An error occurred: {str(e)}")
            log.write(f"ERROR: {str(e)}\n")

if __name__ == "__main__":
    main()
